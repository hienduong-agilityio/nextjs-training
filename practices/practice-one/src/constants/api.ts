export const HTTP_METHODS = {
  GET: 'GET',
  POST: 'POST',
  PUT: 'PUT',
  DELETE: 'DELETE',
};

const ENV = {
  ROOT_API: process.env.NEXT_PUBLIC_SERVER_URL,
  USER_ID: process.env.NEXT_PUBLIC_USER_ID,
};

export const API_URL = {
  PRODUCT: `${ENV.ROOT_API}/product`,
  CART: `${ENV.ROOT_API}/carts`,
};

export const USER_ID = ENV.USER_ID;
