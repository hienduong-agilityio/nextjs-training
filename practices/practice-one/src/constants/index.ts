export * from './api';
export * from './category';
export * from './searchParam';
export * from './router';
