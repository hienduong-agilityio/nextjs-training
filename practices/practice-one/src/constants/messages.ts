// Toast Messages
export const TOAST_MESSAGES = {
  ADD_SUCCESS: 'Product added to cart successfully!',
  ADD_MAX_QUANTITY: 'Unable to add product. Maximum quantity exceeded!',
  ADD_FAILED: 'Failed to add product to cart. Please try again!',
  DELETE_SUCCESS: 'Product removed from cart successfully!',
  DELETE_FAILED: 'Failed to remove product from cart. Please try again!',
  API_ERROR: 'API Error. Please try again!',
};

export const TOAST_TYPES = {
  SUCCESS: 'success',
  ERROR: 'error',
  WARNING: 'warning',
};
