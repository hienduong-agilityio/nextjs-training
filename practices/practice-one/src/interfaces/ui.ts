export interface IInputGroupProps {
  placeholder?: string;
  value?: string;
  buttonText?: string;
  inputName?: string;
  isDisabled?: boolean;
  isLoading?: boolean;
  startIcon?: React.ReactNode;
  customClass?: {
    container?: string;
    inputContainer?: string;
    input?: string;
    button?: string;
  };
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void;
}
