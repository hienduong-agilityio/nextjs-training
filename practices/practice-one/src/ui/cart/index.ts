export * from './CartTable';
export * from './OrderSummary';
export * from './VoucherInput';
export * from './EmptyCart';

export { default as CartItemRow } from './CartItemRow';
