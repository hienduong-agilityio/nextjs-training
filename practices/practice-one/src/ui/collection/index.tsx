export * from './LoadingCategoryGroup';
export * from './LoadingFilterSortBar';
