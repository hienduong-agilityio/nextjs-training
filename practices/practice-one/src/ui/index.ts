export * from './products/index';
export * from './collection/index';
