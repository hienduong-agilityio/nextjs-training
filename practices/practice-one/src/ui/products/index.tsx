export * from './CartAndFavoriteActions';
export * from './ProductImages';
export * from './ProductInfo';
export * from './QuantityControl';
export * from './SocialShare';
export * from './RelatedProduct';
export * from './ProductDetailTabs';
export * from './ProductReview';
export * from './ProductActions';
export * from './ProductLoading';
