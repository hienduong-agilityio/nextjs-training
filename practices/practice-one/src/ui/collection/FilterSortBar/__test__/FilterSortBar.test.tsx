import { render, screen, fireEvent } from '@testing-library/react';
import { useSearchParams, useRouter } from 'next/navigation';

// UI
import { FilterSortBar } from '@/ui';

// Constants
import { SORT_PRODUCT_OPTIONS } from '@/constants';

jest.mock('next/navigation', () => ({
  useSearchParams: jest.fn(),
  useRouter: jest.fn(),
}));

const mockUseSearchParams = useSearchParams as jest.Mock;
const mockUseRouter = useRouter as jest.Mock;

const setupMocks = (searchParams = new URLSearchParams(), push = jest.fn()) => {
  mockUseSearchParams.mockReturnValue(searchParams);
  mockUseRouter.mockReturnValue({ push });

  return { push };
};

const renderFilterSortBar = (props = {}) => {
  const defaultProps = {
    itemCount: 10,
    sortOptions: SORT_PRODUCT_OPTIONS,
    showOptions: ['6', '12', '24'],
    ...props,
  };

  return render(<FilterSortBar {...defaultProps} />);
};

describe('FilterSortBar Component', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly with default props', () => {
    setupMocks();
    const { asFragment } = renderFilterSortBar();

    expect(screen.getByText('10 Items')).toBeInTheDocument();
    expect(screen.getByText('Sort')).toBeInTheDocument();
    expect(screen.getByText('Show')).toBeInTheDocument();
    SORT_PRODUCT_OPTIONS.forEach((option) => {
      expect(screen.getByText(option)).toBeInTheDocument();
    });

    expect(asFragment()).toMatchSnapshot();
  });

  it('handles empty sort and show options gracefully', () => {
    setupMocks();
    renderFilterSortBar({ itemCount: 0, sortOptions: [], showOptions: [] });

    expect(screen.getByText('0 Items')).toBeInTheDocument();
    expect(screen.queryByRole('option')).not.toBeInTheDocument();
  });

  it('calls updateQueryParams when sort option changes', () => {
    const { push } = setupMocks(new URLSearchParams());

    renderFilterSortBar();

    const sortSelect = screen.getByText('Sort')
      .nextSibling as HTMLSelectElement;
    fireEvent.change(sortSelect, {
      target: { value: SORT_PRODUCT_OPTIONS[1] },
    });

    expect(push).toHaveBeenCalledWith(`?sortBy=${SORT_PRODUCT_OPTIONS[1]}`);
  });

  it('calls updateQueryParams when show option changes', () => {
    const { push } = setupMocks(new URLSearchParams());

    renderFilterSortBar();

    const showSelect = screen.getByText('Show')
      .nextSibling as HTMLSelectElement;
    fireEvent.change(showSelect, { target: { value: '12' } });

    expect(push).toHaveBeenCalledWith('?limit=12');
  });

  it('resets pagination when show option changes', () => {
    const initialParams = new URLSearchParams({ page: '2' });
    const { push } = setupMocks(initialParams);

    renderFilterSortBar();

    const showSelect = screen.getByText('Show')
      .nextSibling as HTMLSelectElement;
    fireEvent.change(showSelect, { target: { value: '24' } });

    expect(push).toHaveBeenCalledWith('?limit=24');
    expect(push).not.toHaveBeenCalledWith(expect.stringContaining('page=2'));
  });
});
