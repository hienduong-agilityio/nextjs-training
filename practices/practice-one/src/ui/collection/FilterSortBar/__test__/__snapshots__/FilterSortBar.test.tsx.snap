// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FilterSortBar Component renders correctly with default props 1`] = `
<DocumentFragment>
  <div
    class="bg-secondary-100 p-4 shadow-md md:sticky rounded top-0 z-30 w-full"
  >
    <div
      class="flex flex-col sm:flex-row sm:justify-between sm:items-center space-y-4 sm:space-y-0"
    >
      <div
        class="text-sm w-max text-gray-600"
      >
        10 Items
      </div>
      <div
        class="flex flex-col sm:flex-row sm:items-center sm:space-x-4 space-y-4 sm:space-y-0"
      >
        <div
          class="flex gap-2 items-center"
        >
          <span>
            Sort
          </span>
          <select
            class="border text-sm rounded px-3 py-2 transition duration-200 ease-in-out focus:ring-2 focus:outline-none cursor-pointer border-gray-300 hover:border-gray-400 focus:ring-gray-300 w-auto bg-secondary-100 capitalize"
          >
            <option
              value="name"
            >
              name
            </option>
            <option
              value="price"
            >
              price
            </option>
            <option
              value="rating"
            >
              rating
            </option>
          </select>
        </div>
        <div
          class="flex gap-2 items-center"
        >
          <span>
            Show
          </span>
          <select
            class="border text-sm rounded px-3 py-2 transition duration-200 ease-in-out focus:ring-2 focus:outline-none cursor-pointer border-gray-300 hover:border-gray-400 focus:ring-gray-300 w-auto bg-secondary-100"
          >
            <option
              value="6"
            >
              6
            </option>
            <option
              value="12"
            >
              12
            </option>
            <option
              value="24"
            >
              24
            </option>
          </select>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
